File: shared/BuddyCore/Sources/BuddyCore/Connectivity/ConnectivityBridge.swift
import Foundation

/// The transport-agnostic interface every connectivity adapter implements.
/// `WifiTransport` and `BleTransport` both conform; `ConnectivityBridge`
/// orchestrates them as a failover ladder.
public protocol BuddyTransport: AnyObject {
    /// Start advertising as a host. Calls `onPeerConnected` once a guest
    /// has connected, and `onFrame` for every inbound frame.
    func startHosting(localPeer: Peer) async throws
    func stopHosting()

    /// Start scanning for hosts. Surfaces them via `onHostsChanged`.
    func startScanning(localPeer: Peer) async throws
    func stopScanning()

    /// Connect to a discovered host.
    func connect(to host: DiscoveredPeer) async throws

    /// Send a length-prefixed frame.
    func send(_ frame: Data) async throws

    /// Disconnect cleanly.
    func disconnect()

    /// Live observers — set before calling `startHosting` / `startScanning`.
    var onHostsChanged: (([DiscoveredPeer]) -> Void)? { get set }
    var onPeerConnected: ((Peer) -> Void)? { get set }
    var onFrame: ((Data) -> Void)? { get set }
    var onDisconnected: ((Error?) -> Void)? { get set }
}

/// A peer that the local discovery layer has surfaced but not yet connected
/// to. `peerId` is parsed from the host's advertised metadata.
public struct DiscoveredPeer: Hashable, Sendable, Identifiable {
    public let id: String          // wire id (host:port for Wi-Fi, GATT addr for BLE)
    public let peerId: UUID?
    public let displayName: String
    public let transport: Transport

    public init(id: String, peerId: UUID?, displayName: String, transport: Transport) {
        self.id = id
        self.peerId = peerId
        self.displayName = displayName
        self.transport = transport
    }
}

/// State machine for the failover ladder. The UI's Connect screen listens
/// to `state` and renders accordingly.
public final class ConnectivityBridge {

    public enum State: Equatable, Sendable {
        case idle
        case advertising(via: Transport)
        case scanning(via: Transport)
        case connecting(to: DiscoveredPeer)
        case connected(peer: Peer, via: Transport)
        case failed(reason: String)
    }

    public enum Preference: String, Codable, Sendable {
        case auto, wifiOnly, bleOnly
    }

    public private(set) var state: State = .idle {
        didSet { onStateChanged?(state) }
    }
    public var preference: Preference = .auto

    public var onStateChanged: ((State) -> Void)?
    public var onHostsChanged: (([DiscoveredPeer]) -> Void)?
    public var onFrame: ((Data) -> Void)?

    private let wifi: BuddyTransport
    private let ble: BuddyTransport

    public init(wifi: BuddyTransport, ble: BuddyTransport) {
        self.wifi = wifi
        self.ble = ble
        wireObservers()
    }

    // MARK: - Public API

    public func host(as localPeer: Peer) async throws {
        switch preference {
        case .auto, .wifiOnly:
            do {
                try await wifi.startHosting(localPeer: localPeer)
                state = .advertising(via: .wifi)
                if preference == .auto {
                    // Also advertise on BLE so phones with no Wi-Fi can find us.
                    try? await ble.startHosting(localPeer: localPeer)
                }
                return
            } catch {
                if preference == .wifiOnly { state = .failed(reason: "Wi-Fi advertising failed"); throw error }
            }
            // Fall through to BLE.
            fallthrough
        case .bleOnly:
            try await ble.startHosting(localPeer: localPeer)
            state = .advertising(via: .ble)
        }
    }

    public func scan(as localPeer: Peer) async throws {
        switch preference {
        case .auto, .wifiOnly:
            try await wifi.startScanning(localPeer: localPeer)
            state = .scanning(via: .wifi)
            if preference == .auto {
                try? await ble.startScanning(localPeer: localPeer)
            }
        case .bleOnly:
            try await ble.startScanning(localPeer: localPeer)
            state = .scanning(via: .ble)
        }
    }

    public func connect(to host: DiscoveredPeer) async throws {
        state = .connecting(to: host)
        let t = host.transport == .ble ? ble : wifi
        try await t.connect(to: host)
    }

    public func send(_ frame: Data) async throws {
        switch state {
        case .connected(_, let t):
            let transport = t == .ble ? ble : wifi
            try await transport.send(frame)
        default:
            throw BridgeError.notConnected
        }
    }

    public func disconnect() {
        wifi.disconnect()
        ble.disconnect()
        state = .idle
    }

    // MARK: - Wiring

    private func wireObservers() {
        wifi.onHostsChanged = { [weak self] in self?.aggregateHosts(wifi: $0) }
        ble.onHostsChanged  = { [weak self] in self?.aggregateHosts(ble:  $0) }
        wifi.onFrame = { [weak self] in self?.onFrame?($0) }
        ble.onFrame  = { [weak self] in self?.onFrame?($0) }
        wifi.onPeerConnected = { [weak self] in self?.state = .connected(peer: $0, via: .wifi) }
        ble.onPeerConnected  = { [weak self] in self?.state = .connected(peer: $0, via: .ble) }
        wifi.onDisconnected = { [weak self] _ in self?.state = .idle }
        ble.onDisconnected  = { [weak self] _ in self?.state = .idle }
    }

    private var lastWifi: [DiscoveredPeer] = []
    private var lastBle:  [DiscoveredPeer] = []
    private func aggregateHosts(wifi: [DiscoveredPeer]) {
        lastWifi = wifi
        onHostsChanged?(lastWifi + lastBle)
    }
    private func aggregateHosts(ble: [DiscoveredPeer]) {
        lastBle = ble
        onHostsChanged?(lastWifi + lastBle)
    }

    public enum BridgeError: Swift.Error, Equatable {
        case notConnected
        case noTransportAvailable
    }
}

File: apps/BuddyPlay/android/core/src/main/java/com/americangroupllc/buddyplay/core/connectivity/ConnectivityBridge.kt
package com.americangroupllc.buddyplay.core.connectivity

import com.americangroupllc.buddyplay.core.models.Peer
import com.americangroupllc.buddyplay.core.models.Transport

/**
 * State machine for the failover ladder. The UI's Connect screen listens
 * to [state] and renders accordingly.
 *
 * Mirrors `shared/BuddyCore/Sources/BuddyCore/Connectivity/ConnectivityBridge.swift`.
 */
class ConnectivityBridge(
    private val wifi: BuddyTransport,
    private val ble: BuddyTransport,
) {
    sealed class State {
        object Idle : State()
        data class Advertising(val via: Transport) : State()
        data class Scanning(val via: Transport) : State()
        data class Connecting(val to: DiscoveredPeer) : State()
        data class Connected(val peer: Peer, val via: Transport) : State()
        data class Failed(val reason: String) : State()
    }

    enum class Preference { AUTO, WIFI_ONLY, BLE_ONLY }

    var state: State = State.Idle
        private set(value) { field = value; onStateChanged?.invoke(value) }

    var preference: Preference = Preference.AUTO

    var onStateChanged: ((State) -> Unit)? = null
    var onHostsChanged: ((List<DiscoveredPeer>) -> Unit)? = null
    var onFrame: ((ByteArray) -> Unit)? = null

    private var lastWifi: List<DiscoveredPeer> = emptyList()
    private var lastBle:  List<DiscoveredPeer> = emptyList()

    init { wireObservers() }

    suspend fun host(localPeer: Peer) {
        when (preference) {
            Preference.AUTO, Preference.WIFI_ONLY -> {
                try {
                    wifi.startHosting(localPeer)
                    state = State.Advertising(Transport.WIFI)
                    if (preference == Preference.AUTO) {
                        runCatching { ble.startHosting(localPeer) }
                    }
                    return
                } catch (e: Throwable) {
                    if (preference == Preference.WIFI_ONLY) {
                        state = State.Failed("Wi-Fi advertising failed: ${e.message}")
                        throw e
                    }
                }
                ble.startHosting(localPeer)
                state = State.Advertising(Transport.BLE)
            }
            Preference.BLE_ONLY -> {
                ble.startHosting(localPeer)
                state = State.Advertising(Transport.BLE)
            }
        }
    }

    suspend fun scan(localPeer: Peer) {
        when (preference) {
            Preference.AUTO, Preference.WIFI_ONLY -> {
                wifi.startScanning(localPeer)
                state = State.Scanning(Transport.WIFI)
                if (preference == Preference.AUTO) {
                    runCatching { ble.startScanning(localPeer) }
                }
            }
            Preference.BLE_ONLY -> {
                ble.startScanning(localPeer)
                state = State.Scanning(Transport.BLE)
            }
        }
    }

    suspend fun connect(host: DiscoveredPeer) {
        state = State.Connecting(host)
        val t = if (host.transport == Transport.BLE) ble else wifi
        t.connect(host)
    }

    suspend fun send(frame: ByteArray) {
        val s = state
        if (s !is State.Connected) throw NotConnectedException()
        val t = if (s.via == Transport.BLE) ble else wifi
        t.send(frame)
    }

    fun disconnect() {
        wifi.disconnect()
        ble.disconnect()
        state = State.Idle
    }

    private fun wireObservers() {
        wifi.onHostsChanged = { lastWifi = it; onHostsChanged?.invoke(lastWifi + lastBle) }
        ble.onHostsChanged  = { lastBle  = it; onHostsChanged?.invoke(lastWifi + lastBle) }
        wifi.onFrame = { onFrame?.invoke(it) }
        ble.onFrame  = { onFrame?.invoke(it) }
        wifi.onPeerConnected = { state = State.Connected(it, Transport.WIFI) }
        ble.onPeerConnected  = { state = State.Connected(it, Transport.BLE) }
        wifi.onDisconnected = { state = State.Idle }
        ble.onDisconnected  = { state = State.Idle }
    }

    class NotConnectedException : RuntimeException("not connected")
}
